### ISSUE TITLE
Health checks not relaunching container when EC2 host is terminated.

### ISSUE THREAD

**Issue by richbos** [rancher/rancher#7420]
**Rancher Versions:**
Server: 1.3.1
healthcheck: rancher/healthcheck:v0.2.0
ipsec: rancher/net:holder
network-services: rancher/network-manager:v0.4.0
scheduler: rancher/scheduler:v0.5.1
kubernetes (if applicable):

**Docker Version:**
1.12.3

**OS and where are the hosts located? (cloud, bare metal, etc):**
RancherOS
AWS EC2

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node Rancher server, external RDS DB, x4 EC2 hosts (x2 tagged as eu-west1-goagent + x2 tagged as eu-west1-goserver) 

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
Deploy go-cd server service as a single container via scheduling to named host label eu-west1-goserver.
Terminate the host running the gocd-server service from within the AWS EC2 console.

**Results:** 
Host drops to disconnected status, gocd service drops to unhealthy and remains so, even when host manually deactivated and/or the host is totally deleted from the Rancher UI console.
The service container **only** relaunches in a healthy host when 'Finish Upgrade' is manually instigated.

**Expected:**
Service to relaunch on a healthy host after specified health check period (i.e. x2 instances of 4000ms)

To clarify why we are doing this by terminating a host instance. It is a test of EC2 zone/instance/network outages, we need to be confident that Rancher container deploy/manage is resilient.

![screen shot 2017-01-15 at 18 38 42](https://cloud.githubusercontent.com/assets/1983458/21965279/824de910-db54-11e6-9952-cd8f21733006.png)
![screen shot 2017-01-15 at 18 50 33](https://cloud.githubusercontent.com/assets/1983458/21965280/824e925c-db54-11e6-8c71-56703daff7dd.png)
![screen shot 2017-01-15 at 18 53 54](https://cloud.githubusercontent.com/assets/1983458/21965281/82505c9a-db54-11e6-82b7-3f916bc4f73d.png)

**Comment by richbos** [comment]
To add to this. I have tested outside of AWS EC2 with a local Vagrant/Virtual Box set up this morning and the issue is the same. Kill the underlying host and the container doesn't relaunch on a healthy host, even though it is reported as unhealthy and the host identified as disconnected.

![screen shot 2017-01-16 at 09 23 07](https://cloud.githubusercontent.com/assets/1983458/21977290/705fff58-dbcd-11e6-8538-531d246ae30b.png)

**Comment by vincent99** [comment]
@richbos that doesn't sound like the same issue.  You're service is in the middle of an upgrade and we do not currently reconcile (reschedule) containers that are involved.  Clicking finish upgrade puts the service back into normal active state.

**Comment by richbos** [comment]
Hi, many thanks for the reply.

Yes, it would seem the upgrade process is indeed what is happening, although should this not be automatic? i.e if an underlying host dies and the health check reports an unhealthy container + disconnected host I would expect the container to re-launch on a healthy host?

**Comment by richbos** [comment]
Maybe I didn't explain the issue correctly? Or what my expectations of how Rancher works are....

Basically, if we have a selection of EC2 hosts, and one of them becomes unavailable for whatever reason (we are simulating this by terminating the instance) then I would expect Rancher to relocate the containers which were running on the failed host to another healthy host. However this isn't happening for us, not in EC2 or in a localised test environment using VirtualBox with Vagrant initialised hosts.

**Comment by richbos** [comment]
Just to further clarify, the service is not initially in an upgrade state, and does not change to an upgrade state UNTIL the underlying host is removed, it then changes (to the upgrade state) and requires manual intervention to restart/relocate to a healthy host. This is not expected behaviour.

**Comment by dayglo** [comment]
I'm evaluating rancher right now and am experiencing what i think is the same issue. i kill my host and none of my containers are restarted anywhere else.  Isn't the whole idea of cattle to make sure my stacks are always running?

**Comment by richbos** [comment]
@vincent99 Hi, just wondering if there might be any movement imminent on this issue?

Unfortunately we need a solution for our development route so if not we may have to divert efforts to evaluating AWS ECS in place of Rancher.

**Comment by vincent99** [comment]
@richbos Rescheduling because of a healthcheck failure does not trigger an 'upgrade' and doesn't put the service in the state your screenshot suggests.  It still seems like you're doing an Upgrade at some point and never clicking to finish it.

To see a container move over you need:

- A service starting in **active** state (which would *not* have a "Finish Upgrade" action)
- A healthcheck on that service which will pass while the service is up (so it leaves `initializing` health state) and then fail when it's down (that's the easy part).
-A trigger to make the healthcheck fail.  Note that just marking a host in Rancher as 'inactive' only prevents new containers from being scheduled to it, it doesn't evacuate existing ones (that will eventually be a separate action).  Stopping the instance in EC2 is a straightforward way (since your screenshot shows `disconnected` you have this part working)
- Another eligible host to put the service container on.  For example the other host(s) may not be eligible if the service is defined to only go on that particular host, or to require a label only the host has, or require not having a label every other host has, or the ports it publishes are in use, etc.

**Comment by richbos** [comment]
Hi @vincent99 - I understand what you are saying but we weren't doing an upgrade, the container was in an active state prior to the host being taken down, this was the same in both of the environments (EC2  and VirtualBox). The host label specified was tagged to x2 hosts.

We tested this many, many times, both in EC2 over a period of several days, however, I have just re-tested now and the container relocated when the underlying host was stopped, so it would appear that without any intervention the problem has resolved itself, which is even more baffling.

Here are grabs detailing the configuration.

![screen shot 2017-01-18 at 17 09 39](https://cloud.githubusercontent.com/assets/1983458/22074932/10dbce32-dda2-11e6-9515-9aebb8686e5a.png)

![screen shot 2017-01-18 at 17 09 59](https://cloud.githubusercontent.com/assets/1983458/22074946/19080f44-dda2-11e6-9a43-ae7a8c541fee.png)

![screen shot 2017-01-18 at 17 10 29](https://cloud.githubusercontent.com/assets/1983458/22074960/208694fc-dda2-11e6-834b-c836d3041de3.png)

![screen shot 2017-01-18 at 17 13 54](https://cloud.githubusercontent.com/assets/1983458/22074970/27399c36-dda2-11e6-9610-1bdf20cd6ac0.png)

....and finally, the result of the failed host from today as we would have expected, i.e with the container automatically relaunched on a healthy host.

![screen shot 2017-01-18 at 17 18 49](https://cloud.githubusercontent.com/assets/1983458/22075021/52786530-dda2-11e6-9e80-eda78d781d68.png)

**Comment by richbos** [comment]
Ok @vincent99 - My colleague thinks this may be down to a need for 3 hosts, i.e in order to decide on 'split brain' state?

For example, re this - 

_“Rancher exhibits the characteristics of an active-passive HA deployment, so the number of host failures that may be tolerated is computed as `ceil(N / 2) – 1`, where N is the number of Rancher nodes. In simpler terms, a 3-node deployment survives 1 host failure, a 5-node deployment survives 2 host failures, and so on."_

Would that be it? Do we require x3 hosts per service?

RB

**Comment by vincent99** [comment]
I also hear what you're saying but the only way you get the screenshot showing "Finish Upgrade" is if the service is in Upgraded state, and the only way that happens is if a human triggers it.  Maybe it was someone else with access to the environment?

The HA setup (if you use it) is active/active and has that (n/2)-1 property like any distributed system.   But this is not relevant to your problem unless these hosts are *also* the ones running the `rancher/server`.  All the hosts registered to an environment can fail and the orchestration system is still fine because it's not running on those hosts.

There is a small benefit to having >= 4 hosts in an environment, in that it allows the optimal health check setup.  Health checks for a container are performed on (up to) 3 hosts, separate from the one one the container lives on.  So if you have 4+ hosts you get checks from 3 independent hosts.  With <= 3 one of the checks goes on the same host as the container is on, and with =1 host it's *only* the host checking itself.

**Comment by richbos** [comment]
Hi @vincent99, there is nobody else has access to our environments, not in ec2 and certainly not locally in my home office.
We shall re-test again tomorrow and post relevant before and after grabs to verify the upgrade state as described.

The issue as we see it then is that for some reason that with our configuration the service seems to be entering the upgrade state as a result of the underlying host being removed. This is what we need to clarify is happening and I guess if we can reproduce that then we can take if from there.

I shall report back tomorrow.

**Comment by richbos** [comment]
Hi @vincent99, many thanks for your attention and help with this. 

We have carried out extra tests and yes, it was down to our misunderstanding of the Rancher service and container launch process.
When we launch our container service we need it to run on certain hosts, as such we do not start it but 'upgrade' for host tags using scheduling, we then click 'upgrade'.
What we didn't realise was that even as the service has become active there is another manual step, i.e. after clicking upgrade you then have to wait and click 'finish upgrade'. We didn't realise this, we thought as it was active, then it was active.

So, the issue for us has changed. Ideally we would like to be able to select scheduled host tags at container/service launch. We don't really see this as an 'upgrade', but more of a 'configuration' option, it would also streamline the process for us. Do you think this might be possible as a feature request? Or is there logic to the way the process currently is and it is something we would need to live with?

**Comment by vincent99** [comment]
Are you deploying things from catalog, and turning off start on create so you can change them before they first run?  That's basically https://github.com/rancher/rancher/issues/2852 , which we'll have a solution for eventually..

**Comment by richbos** [comment]
Hi @vincent99 , yes, that is exactly it...! Great to hear you are working on a solution, that would be perfect for our needs.

Thank you again for your advice and patience.

**Comment by aemneina** [comment]
related tickets for this have been fixed and closed. Please file a new ticket if issue persists in rancher 1.5.9+.

https://github.com/rancher/rancher/issues/3239
